### PR TITLE
Print stacktrace on segfaults

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

This seems to work on the machines that I have tested (Wheeler, macOS, CI container). It would have saved me so much time and sanity over the years. The printed stacktraces are improved a lot by #3907.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

**Comment by nilsvu** [comment]
@markscheel turns out this does print useful information for "Unrecognized PUP::able::PUP_ID" segfaults. Even without the prettier stack trace from #3907:

```
Shortened stack trace is:
3   _PyDomain.so                        0x00000001073d642c _ZN12_GLOBAL__N_123segfault_signal_handlerEi + 176
4   libsystem_platform.dylib            0x000000018d616a84 _sigtramp + 56
5   _PyDomain.so                        0x0000000107577b40 _Z9LrtsAbortPKc + 48
6   _PyDomain.so                        0x0000000107577b40 _Z9LrtsAbortPKc + 48
7   _PyDomain.so                        0x0000000107577b10 _Z9LrtsAbortPKc + 0
8   _PyDomain.so                        0x0000000107573b1c CmiSyncBroadcastAllFn + 0
9   _PyDomain.so                        0x00000001075598c4 _Z15PUP_getRegEntryRKN3PUP4able6PUP_IDE + 160
10  _PyDomain.so                        0x0000000107559adc _ZN3PUP2er6objectEPPNS_4ableE + 316
11  _PyDomain.so                        0x00000001070c0d30 _ZN6domainorERN3PUP2erERPNS_17CoordinateMapBaseIN5Frame12BlockLogicalENS4_4GridELm3EEE + 44
12  _PyDomain.so                        0x00000001070c0cd8 _ZN3PUP3pupIN6domain17CoordinateMapBaseIN5Frame12BlockLogicalENS3_4GridELm3EEELDn0EEEvRNS_2erERNSt3__110unique_ptrIT_NS9_14default_deleteISB_EEEE + 60
```

There's `CoordinateMapBase<Frame::BlockLogical, Frame::Grid>` in the stack trace. Of course with the demangling in #3907 this is way easier to read.

File: tests/Unit/ErrorHandling/Test_SegfaultHandler.cpp [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+// [[OutputRegex, Segmentation fault!]]
+SPECTRE_TEST_CASE("Unit.ErrorHandling.SegfaultHandler",
+                  "[ErrorHandling][Unit]") {
+  ERROR_TEST();
+  enable_segfault_handler();
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wold-style-cast"
+  // NOLINTNEXTLINE
+  int* foo = (int*)-1;  // make a bad pointer
```
Comment (wthrowe) [inline tests/Unit/ErrorHandling/Test_SegfaultHandler.cpp]: Why an old-style cast?

File: tests/Unit/ErrorHandling/Test_SegfaultHandler.cpp [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+#pragma GCC diagnostic pop
+  printf("%d\n", *foo);  // causes segfault
+  CHECK(true);
```
Comment (wthrowe) [inline tests/Unit/ErrorHandling/Test_SegfaultHandler.cpp]: Drop line.  If we get here the test has failed.

File: tests/Unit/ErrorHandling/Test_SegfaultHandler.cpp [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+SPECTRE_TEST_CASE("Unit.ErrorHandling.SegfaultHandler",
+                  "[ErrorHandling][Unit]") {
+  ERROR_TEST();
+  enable_segfault_handler();
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wold-style-cast"
+  // NOLINTNEXTLINE
+  int* foo = (int*)-1;  // make a bad pointer
+#pragma GCC diagnostic pop
+  printf("%d\n", *foo);  // causes segfault
```
Comment (wthrowe) [inline tests/Unit/ErrorHandling/Test_SegfaultHandler.cpp]: We may need to do something more clever to trick the optimizer, but we don't need to deal with that unless it becomes an issue.

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: tests/Unit/ErrorHandling/Test_SegfaultHandler.cpp (hunk 1 above)
Comment (nilsvu) [reply tests/Unit/ErrorHandling/Test_SegfaultHandler.cpp]: This doesn't work reliably anyway. Do you know of a reliable way of triggering a segfault but not a bus error or something like that?

Comment (nilsvu) [reply tests/Unit/ErrorHandling/Test_SegfaultHandler.cpp]: Actually nvm, I went with raising the signal manually.

**Review by wthrowe — APPROVED** [review]

**Review by kidder — APPROVED** [review]

**Pull request merged**